### PR TITLE
doc: tor.md updates

### LINKED ISSUE #19924: Onion address in Bitcoin Core Wallet
**Is your feature request related to a problem? Please describe.**

I am not sure if there is a RPC method available to directly get the onion address used so I checked it in `localaddresses` returned by `getnetworkinfo` https://developer.bitcoin.org/reference/rpc/getnetworkinfo.html

It was also mentioned in debug.log file:

image

**Describe the solution you'd like**

I would love to see an onion address and maybe qr code in the wallet if it is listening on Tor similar to [ABCore](https://github.com/greenaddress/abcore/blob/cc629a854f3083cf4bd1a743de80a254c72aa8ea/app/src/main/java/com/greenaddress/abcore/RPCIntentService.java#L134):

image

**Describe alternatives you've considered**

Check the onion address in debug.log or `localaddresses` returned for `getnetworkinfo`

**Additional context**

Linked to https://github.com/bitcoin/bitcoin/issues/19923 because if we add such option to check the onion address, it will have to be added in the [Tor documentation](https://github.com/bitcoin/bitcoin/blob/master/doc/tor.md).

### PR DESCRIPTION
**Pull request by jonatack**
It looks like `doc/tor.md` could use some updates and improvements, not only for Tor v3, but also for setting multiple addresses with `-externalip` (see the conversation from http://www.erisian.com.au/bitcoin-core-dev/log-2020-09-16.html#l-39), how to see information about your Tor config via Bitcoin Core, and other improvements.

Closes #19924.

### CODE REVIEW

**Review by RiccardoMasutti — APPROVED** [review]
ACK. Seems good to me.

**Comment by Saibato** [comment]
Pls feel free to grab from  https://github.com/bitcoin/bitcoin/commit/654dc04b67a44199352a1b25494c9d617cd090bc

Then I can strip this from #19358 and rename that and re push
As.
**torcontrol :do not override with default proxy settings in hidden service, if the user has chosen otherwise according with the wording in the docs .**

Since we until now notoriously in disrespect override the users wished settings with the defaults in torcontrol.

PR  #19358 should be applied or tor.md doc changed to reflect that footgun.

**Comment by jonatack** [comment]
@Saibato thanks, I'll look at testing and pulling in that commit. In any case I need to verify and improve this draft. @RiccardoMasutti thanks.

File: doc/tor.md
Comment (hebasto) [inline doc/tor.md:57]: When a hidden service is created via tor-control, `/var/lib/tor/bitcoin-service/hostname` is not created (credits to @vasild).

Comment (vasild) [reply doc/tor.md:57]: But this new text reads:

> Given the above configuration ...

and just a few lines above it says how to configure the hidden service via `torrc`:

> HiddenServiceDir /var/lib/tor/bitcoin-service/
> HiddenServicePort 8333 127.0.0.1:8333
> HiddenServicePort 18333 127.0.0.1:18333

So it is ok to mention `/var/lib/tor/bitcoin-service/hostname`.

I would suggest to drop the last line `HiddenServicePort 18333 127.0.0.1:18333` because it is for testnet and also to add `HiddenServiceVersion 3` so that the tor daemon creates a v3 service:

```
HiddenServiceDir /var/lib/tor/bitcoin-service/
HiddenServiceVersion 3
HiddenServicePort 8333 127.0.0.1:8333
```

**Review by vasild — APPROVED** [review]
ACK a326956

**Comment by vasild** [comment]
Maybe split the white-space changes in a separate commit to ease further reviewers.

File: doc/tor.md
Comment (hebasto) [reply doc/tor.md:57]: Does `bitcoind` configure its onion service via `torrc`?

**Comment by prayank23** [comment]
Examples to run Bitcoin as onion service, use of tor bridges and safe ways to connect over tor: https://github.com/bitcoin/bitcoin/issues/19923

We can mention "How to check onion address"? : https://github.com/bitcoin/bitcoin/issues/19924

**Review by ghost — COMMENTED** [review]
Few things that we can add in the updates:

Examples to run Bitcoin as onion service, use of tor bridges and safe ways to connect over tor: https://github.com/bitcoin/bitcoin/issues/19923

We can mention "How to check onion address"? : https://github.com/bitcoin/bitcoin/issues/19924

**Comment by Willtech** [comment]
> Few things that we can add in the updates:
> 
> Examples to run Bitcoin as onion service, use of tor bridges and safe ways to connect over tor: #19923
> 
> We can mention "How to check onion address"? : #19924

I have replied on each FR where it is appropriate.

**Comment by jonatack** [comment]
Thanks everyone for the feedback. There are quite a few updates to do to bring this doc up-to-date. Will bring it out of draft status soon.

**Comment by Rspigler** [comment]
Thanks for taking this on! My concern is that the doc doesn't really help the most common end users.  

> Connecting to Tor's control socket API requires one of two authentication methods to be configured. It also requires the control socket to be enabled, e.g. put ControlPort 9051 in torrc config file. For cookie authentication the user running bitcoind must have read access to the CookieAuthFile specified in Tor configuration.

This is nonsense/scary to most people. This ultimately means that most users either don't end up enabling Hidden Services, or have to go to other websites/tutorials to do so (which could be giving them wrong information).  I think our docs could do a much better job explaining to the layperson how to do so.

**Comment by jonatack** [comment]
I have a number of updates written that I haven't yet pushed. My concern is that they are changing too much and will discourage review and not be merged. It's on my list to scale them down and push soon.

**Comment by prayank23** [comment]
> I think our docs could do a much better job explaining to the layperson how to do so.

Agree. I think adding screenshots and examples could help.

**Comment by jonatack** [comment]
> I think our docs could do a much better job explaining to the layperson how to do so.

I don't know if a tutorial for non-technical people is the kind of doc that should be in the source code repository, but it is indeed beyond the scope of what I planned to propose (which I'm scaling down to be reviewable).

**Comment by Rspigler** [comment]
@jonatack That makes sense.  Maybe that's something I'll work on updating for the BitcoinWiki after the recent tor/GUI changes get merged.

**Comment by jonatack** [comment]
Updated. If everyone can review today, maybe this can still be in 0.21 as rc4 is being finalised. 

It may be easiest to review each commit separately. Happy to drop the last commit if it's too much to review.

**Comment by jonatack** [comment]
I also recommend viewing the diff by clicking on the "Display the rich diff" button at top right.

File: doc/tor.md
Comment (vasild) [inline doc/tor.md:112]: nit: it is not proper to say "since Prog1 version X, Prog2 does this...". It should be "Since Bitcoin Core version 1.2.3 we make use of...".

I guess it is ok to just drop "Since Tor version 0.2.7.1, ".

File: doc/tor.md
Comment (vasild) [inline doc/tor.md:133]: Just suggest `chmod -R 777 .` - it always helps with filesystem permissions problems.

File: doc/tor.md
Comment (vasild) [inline doc/tor.md:144]: Hmm, https://tb-manual.torproject.org is the Tor _Browser_ manual. Would https://www.torproject.org/docs/tor-manual.html.en be more appropriate?

Comment (Saibato) [reply doc/tor.md:144]: > Hmm, 

or
https://2019.www.torproject.org/docs/tor-onion-service.html.en

**Comment by Saibato** [comment]
@jonatack 
for some config and setup hints to add u could add,  u might want to look into
https://github.com/ElementsProject/lightning/blob/master/doc/TOR.md

File: doc/tor.md
Comment (Saibato) [reply doc/tor.md:133]: suggstetion i.e from https://github.com/ElementsProject/lightning/blob/master/doc/TOR.md
```
---
You also need to make your user a member of the Tor group. "Your user" here is whatever user will run lightningd. On Debian-derived systems, the Tor group will most likely be debian-tor. You can try listing all groups with the below command, and check for a debian-tor or tor groupname.

getent group | cut -d: -f1 | sort

Alternately, you could check the group of the cookie file directly. Usually, on most Linux systems, that would be /run/tor/control.authcookie:

stat -c '%G' /run/tor/control.authcookie

Once you have determined the ${TORGROUP} and selected the ${LIGHTNINGUSER} that will run lightningd, run this as root:

usermod -a -G ${TORGROUP} ${LIGHTNINGUSER}

Then restart the computer (logging out and logging in again should also work). Confirm that ${LIGHTNINGUSER} is in ${TORGROUP} by running the groups command as ${LIGHTNINGUSER} and checking ${TORGROUP} is listed.

If the /run/tor/control.authcookie exists in your system, then log in as the user that will run lightningd and check this command:

cat /run/tor/control.authcookie > /dev/null
---
```

File: doc/tor.md
Comment (Rspigler) [inline doc/tor.md:121]: "it can configured" -> "it can be configured"

Comment (jonatack) [reply doc/tor.md]: good eye, fixed

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md]: thanks, done

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md]: rewrote the section based on this, thanks @Saibato

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md]: dropped, thanks

**Comment by jonatack** [comment]
> for some config and setup hints to add u could add, u might want to look into
> [https://github.com/ElementsProject/lightning/blob/master/doc/TOR.md](https://github.com/ElementsProject/lightning/blob/master/doc/TOR.md?rgh-link-date=2020-12-16T16%3A47%3A02Z)
> pls also note that the Torcontrol and onlynet bug fix PR;s are not merged in 0.21
> so maybe some hint her fro the user to be not surprised by unexpected behavior might be a good idea.
> ref. [654dc04](https://github.com/bitcoin/bitcoin/commit/654dc04b67a44199352a1b25494c9d617cd090bc)

@Saibato I rewrote the section 3 commit with that information, and cherry-picked your commit into this PR with a few fixups. Let me know.

**Comment by jonatack** [comment]
> Thanks for taking this on! My concern is that the doc doesn't really help the most common end users.
> 
> > Connecting to Tor's control socket API requires one of two authentication methods to be configured. It also requires the control socket to be enabled, e.g. put ControlPort 9051 in torrc config file. For cookie authentication the user running bitcoind must have read access to the CookieAuthFile specified in Tor configuration.
> 
> This is nonsense/scary to most people. This ultimately means that most users either don't end up enabling Hidden Services, or have to go to other websites/tutorials to do so (which could be giving them wrong information). I think our docs could do a much better job explaining to the layperson how to do so.

@Rspigler in the end, I rewrote it in the last push with info from the c-lightning doc @Saibato suggested. Please have a look if you can.

**Review by Rspigler — APPROVED** [review]
Looks great!

**Comment by jonatack** [comment]
@Rspigler Great! Could you add a line to your comment, that says: `ACK 39858289583cc11fdcefbc8e8716ff09855049a7`

(That's the official approval format; thanks!)

File: doc/tor.md
Comment (Rspigler) [inline doc/tor.md:52]: Does this mean `onlynet=ipv4` `onion=0` does not disable onion connections? This seems confusing.

`-noonion` or `onion=0` is always needed to explicitly disable outbound onion access, as described in the `-proxy` and `onion` section?

Comment (Rspigler) [reply doc/tor.md:52]: (I missed this last commit, glad you asked for the clarification.  I ACKd 42e61805734c36da7a11586aabc7fd5742dba267)

Comment (Saibato) [reply doc/tor.md:52]: > Does this mean onlynet=ipv4 onion=0 does not disable onion connections? This seems confusing.

Yup, a default fallback outbound pure .onion not IP name resolving proxy just for onion address will be created by default even without -proxy definition  if onion!=0 and regardless of -onlynet= if -listenonion=1 ( default) if the inbound Tor onion service could be created successfully by the ever  polling torcontroller in bitcoind   (that is your  advertised Tor addr created by bitcoind independent of what onion services defined in torrc ) The idea was, (i guess)  if that is successfully to create also an outbound proxy to reach Tor onion nodes.

So suppose u have a running bitcoind daemon and edit at some point just torrc to allow the control port or start the Tor service, bitcoind does if -listeonion=1 permanent poll for such changes and instantly creates on the fly an inbound onion address and an outbound proxy.  So if u want to be sure that the outbound default proxy is disabled since u want i.e. just ipv4 set -noonion since -onlynet=ipv4 wont do that and please note the inbound Tor is created regardless of that if -listenonion=1 what is the default.  

If that is a desired way to act depends, fixes or changes to that are not merged yet and at least that's the way it is and its now documented so that users are aware of that.

File: doc/tor.md
Comment (vasild) [inline doc/tor.md:24]: What does it mean "If SOCKS5 is selected"? How does one select SOCKS5 or what other options are there to select from?

I think this should be removed:

```suggestion
      -proxy=ip:port  Set a proxy server. This proxy
```

as `-proxy=` only supports SOCKS5 proxies.

File: doc/tor.md
Comment (Saibato) [reply doc/tor.md:52]: @jonasnick  did edit the comment text to be more precise, so please review ur thumbs up ;-)

File: doc/tor.md
Comment (vasild) [inline doc/tor.md:52]: Huh! :-( This is indeed a strange behavior, but I think something like the following belongs to `bitcoind -help` rather than `doc/tor.md`:

> If -onlynet= is used with other networks and -onlynet=onion is not among them, but -onion= or -proxy is provided then outbound onion connections will still be made! Use -noonion or -onion=0 to disable outbound onion connections in this case.

Given that this is a guide on how to enable and configure Tor, not how to disable it, I would suggest to drop the last part and leave it to just:

```
-onlynet=onion  Make outgoing connections only to .onion addresses. Incoming
	        connections are not affected by this option.
```

File: doc/tor.md
Comment (vasild) [inline doc/tor.md:209]: I think this is excessive and should be removed.

* the "sort" in the getent command is unneeded and on my computer the entire thing prints:

```
operator
_tor
You have new mail in /home/toolame/mailbox
```

so, if I am a dummy user, how do I know which is the tor group?

* stat gives me error:
```
$ stat -c '%G' /run/tor/control.authcookie
stat: illegal option -- c
usage: stat [-FLnq] [-f format | -l | -r | -s | -x] [-t timefmt] [file|handle ...]
```

This doc shouldn't be targeted for users who lack basic knowledge of how filesystem permissions work.

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md:24]: I can drop this, but it was existing before this PR.

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md:52]: That would be removing needed information that was present before this PR!

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md:209]: Are you using a Debian-based system?

Comment (jonatack) [reply doc/tor.md:209]: Dropped the commit.

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md:52]: Dropped the commit completely.

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md:24]: Dropping this commit.

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md:52]: Dropping this commit.

**Comment by jonatack** [comment]
Dropped the last two commits that were attracting comments to reduce the scope to have this merged in time for rc4.

Can everyone please ACK?

**Comment by laanwj** [comment]
ACK a34eceb4cc054b4233e7321de927e8a7a2146301

File: doc/tor.md
Comment (Saibato) [reply doc/tor.md]: @jonatack  i am almost certain there will be an rc5 and some backport's ping @luke-jr, so no time to rush here and have another release that is not precise in what the options and Tor-controller does.

Do u need a detailed descriiption and  log to show the real behavior?

**Comment by vasild** [comment]
ACK a34eceb

File: doc/tor.md
Comment (Saibato) [reply doc/tor.md]: @Rspigler tyi, only explicit -nonion or -onion=0 can prevent outbound onion connections, solely -onllynet=pv4/ipv6 is not sufficient to disable that.

that what is said in the PR now dropped.

```
Please use -noonion or -onion=0 if you want to be sure to have no
outbound onion connections over the default proxy or your defined proxy..
```

https://github.com/bitcoin/bitcoin/commit/654dc04b67a44199352a1b25494c9d617cd090bc#diff-a33b1b15efbc14c65460ef7a883b530ca52144a4f06642e4a188351d959f9894R42

Comment (jonatack) [reply doc/tor.md]: @Saibato I might propose the dropped commits, including yours, in a new PR and make some of the changes suggested by @vasild.

File: doc/tor.md
Comment (Saibato) [reply doc/tor.md]: @vasild 
> -onlynet=onion  Make outgoing connections only to .onion addresses. Incoming
        connections are not affected by this option.

I guess its hard to grasp and i was :woman_facepalming:  too when i stumbled about, but what we believed the code does from src comments or doc and almost any post i am aware about bitcoin and Tor and social media is
not what really happens when assuming option do what they say and u follow the docs.

only -noonion or -onion=0 can prevent that, onlynet= is buggy since  #7553 (2016)
Similar but also fact is that although seerders are not called directly in IP if the DNS returns are random that is not the case when bootstrap over Tor here the seeders have full control and can be sure ( or ther MITM'S or proxys)  that the first entry in the DNS return a exit gets will be called since socks5 works that way all seeders will be called in a distinct sequence over exist and then ADDR gossip that is highly disturbing give the timing correlation bogus Tor relays and exits can do.
Tacking in account that all that happens and wrong configured at bootstrap at least once, *pure* Tor nodes are quite f***ed if they not aware.

File: doc/tor.md
Comment (Rspigler) [reply doc/tor.md]: @Saibato thanks for the explanation/edits.  Although now dropped, would ACK in a new PR

**Comment by Rspigler** [comment]
Post-merge ACK a34eceb4cc054b4233e7321de927e8a7a2146301

**Comment by Willtech** [comment]
FYI: There is a reviewed tutorial here for using Bitcoin Core with Tor that seems to have a permalink, albeit the tutorial deliberately steps over many things in Tor that a user could benefit from having a detailed explanation of, such as participation in the tor network, https://bitcoin.stackexchange.com/q/70069/75001

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md]: > @Saibato thanks for the explanation/edits. Although now dropped, would ACK in a new PR

Done in #20757, please add suggestions or ACKs there.

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md]: Brought back the commits and added also to the -onlynet help in #20757, please add suggestions or ACKs there.

File: doc/tor.md
Comment (jonatack) [reply doc/tor.md]: Revived without the `sort` and with a mention that the command is for Debian in #20757.

File: doc/tor.md
Comment (Willtech) [reply doc/tor.md]: https://github.com/bitcoin/bitcoin/pull/19961#discussion_r492980674
Even given the above configuration, you make a point to consider we are talking about the operation of the parameter.


If the hidden service is created via tor-control and you need to discover it please check you debug.log for lines similar to the following, it should automatically be added as AddLocal
```
2020-08-11T13:42:26Z tor: Got service ID {random_service}, advertising service {random_service}.onion:8333
2020-08-11T13:42:26Z AddLocal({random_service}.onion:8333,4)
```

**Review by felipsoarez — COMMENTED** [review]
ACK a34eceb

**Pull request merged**
